src/Interpreters/ExpressionActions.h

```diff
@@ -28,13 +31,138 @@ using ArrayJoinActionPtr = std::shared_ptr<ArrayJoinAction>;
+/// 1. it's not a function or alias node
+/// 2. it doesn't have any parent.
+/// 3. it is used as a output attribute of this ExpressionActions.
+/// 4. it's the first node of one of its parent, and the parent's
+///    enable_lazy_execution_for_first_argument = false.
+/// 5. it's a common descendant of different function arguments of short-circuit functions
+/// 6. it's parent could not be executed lazily and the parent is not a short-circuit function.
+class ExpressionShortCircuitExecuteController
+{
+public:
+    struct ProfileData
+    {
+        UInt64 sample_rows = 0;
+        UInt64 selected_rows = 0;
+        UInt64 eclapsed = 0;
+    };
+
+    enum ProfileType
+    {
+        NOT_PROFILE = 0,
+        PROFILE_ECLAPSED = 1,
+        PROFILE_SELECTIVITY = 2,
+    };
+
+    explicit ExpressionShortCircuitExecuteController(
+        const ActionsDAGPtr & actions_dag_,
+        const ExpressionActionsSettings & settings);
+
+    ExpressionShortCircuitExecuteController(const ExpressionShortCircuitExecuteController & b)
+        : actions_dag(b.actions_dag)
+        , short_circuit_function_evaluation(b.short_circuit_function_evaluation)
+        , enable_adaptive_reorder_arguments(b.enable_adaptive_reorder_arguments)
+        , sampled_rows(b.sampled_rows.load())
+        , finished_adaptive_reorder_arguements(b.finished_adaptive_reorder_arguements.load())
+        , short_circuit_infos(b.short_circuit_infos)
+        , has_reorderable_short_circuit_functions(b.has_reorderable_short_circuit_functions)
+    {}
+
+    ExpressionShortCircuitExecuteController & operator=(const ExpressionShortCircuitExecuteController & b)
+    {
+        actions_dag = b.actions_dag;
+        short_circuit_function_evaluation = b.short_circuit_function_evaluation;
+        enable_adaptive_reorder_arguments = b.enable_adaptive_reorder_arguments;
+        sampled_rows = b.sampled_rows.load();
+        finished_adaptive_reorder_arguements = b.finished_adaptive_reorder_arguements.load();
+        short_circuit_infos = b.short_circuit_infos;
+        has_reorderable_short_circuit_functions = b.has_reorderable_short_circuit_functions;
+        return *this;
+    }
+
+    bool couldLazyExecuted(const ActionsDAG::Node * node);
+    int  needProfile(const ActionsDAG::Node * node);
+    void addNodeShortCircuitProfile(const ActionsDAG::Node * node, const ProfileData & profile_data_);
+    void tryReorderShortCircuitFunctionsAguments(size_t num_rows);
+    std::vector<size_t> getReorderedArgumentsPosition(const ActionsDAG::Node * node);
+    size_t needSampleRows() const;
+    inline bool hasReorderableShortCircuitFunctions() const { return has_reorderable_short_circuit_functions; }
+
+private:
+    struct ShortCircuitInfo
+    {
+        ProfileData profile_data;
+        bool is_lazy_executed = false;
+        std::vector<size_t> arguments_position;
+        bool is_short_circuit_function_child = false;
+    };
+
+    std::shared_mutex mutex;
```

Comment:
Do I understand right that we need a mutex here because we can execute actions on different blocks in parallel?

Reply:
Yes